### PR TITLE
TFP: full noam stack — ANP + T_max=150 + physics features

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The noam branch's ANP cross-attention decoder gave -58.9% improvement on TandemFoil. TFP has been locked at T_max=10 (every deviation diverged), but that lock was established with a stripped-down config. noam runs T_max=150 stably with richer physics features, larger architecture, and Lion+Lookahead+compile. The hypothesis: TFP's divergence at T_max>10 is an artifact of the minimal config, not a fundamental loss-landscape constraint. With the full noam stack, T_max=150 should be stable — and the combination of ANP decoder + wake/panel physics + 96 slices may produce a dramatic drop in val_primary/field_mse.

Three trials are designed to isolate the contribution of physics features vs. ANP vs. architecture:
- Trial 1: Full noam stack (everything)
- Trial 2: ANP + T_max=150 + arch, no extra physics features
- Trial 3: Architecture + T_max=150 only (no ANP, no extra physics)

Reference: noam ANP decoder result on TandemFoil: -58.9% on val_primary/field_mse vs. radford baseline.

## Instructions

**CRITICAL — Tandem debug discipline:** Before running all trials, run Trial 1 for exactly 3 epochs as a smoke test:

```bash
cd target/icml2026 && python train.py --dataset tandem_foil_set --optimizer lion --lr 1.25e-4 --cosine-t-max 150 --grad-clip 0.5 --weight-decay 1e-2 --ema-decay 0.999 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --model-slices 96 --enable-fourier --anp-srf --asinh-pressure --asinh-scale 0.75 --residual-prediction --surface-refine --enable-te-coord-frame --enable-wake-deficit --enable-wake-angle --enable-cp-panel --enable-cp-panel-tandem-only --cp-panel-scale 0.1 --enable-vortex-panel-velocity --vortex-panel-scale 0.1 --vortex-panel-n 64 --re-stratified-sampling --use-lookahead --compile-model --epochs 3 --wandb_group tfp-noam-stack
```

If the smoke test crashes or produces NaN losses at any epoch, **stop immediately and report the error in a PR comment**. Do not proceed to the full trials.

If the smoke test passes (3 epochs complete with finite losses), run all 3 trials in parallel. Use `--wandb_group tfp-noam-stack` on all runs.

---

**Trial 1 — Full noam stack (highest priority):**
```bash
cd target/icml2026 && python train.py --dataset tandem_foil_set --optimizer lion --lr 1.25e-4 --cosine-t-max 150 --grad-clip 0.5 --weight-decay 1e-2 --ema-decay 0.999 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --model-slices 96 --enable-fourier --anp-srf --asinh-pressure --asinh-scale 0.75 --residual-prediction --surface-refine --enable-te-coord-frame --enable-wake-deficit --enable-wake-angle --enable-cp-panel --enable-cp-panel-tandem-only --cp-panel-scale 0.1 --enable-vortex-panel-velocity --vortex-panel-scale 0.1 --vortex-panel-n 64 --re-stratified-sampling --use-lookahead --compile-model --epochs 999 --wandb_group tfp-noam-stack
```

**Trial 2 — ANP + T_max=150 + 96 slices, no extra physics:**
```bash
cd target/icml2026 && python train.py --dataset tandem_foil_set --optimizer lion --lr 1.25e-4 --cosine-t-max 150 --grad-clip 0.5 --weight-decay 1e-2 --ema-decay 0.999 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --model-slices 96 --enable-fourier --anp-srf --use-lookahead --compile-model --epochs 999 --wandb_group tfp-noam-stack
```

**Trial 3 — Architecture + T_max=150, no ANP, no extra physics (ablation):**
```bash
cd target/icml2026 && python train.py --dataset tandem_foil_set --optimizer lion --lr 1.25e-4 --cosine-t-max 150 --grad-clip 0.5 --weight-decay 1e-2 --ema-decay 0.999 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --model-slices 96 --enable-fourier --use-lookahead --compile-model --epochs 999 --wandb_group tfp-noam-stack
```

**Reporting:** For each trial, report both `val_primary/field_mse` AND `test_primary/field_mse` from the best validation checkpoint. This is paper-facing if any trial beats baseline.

Budget: ~6 hours total wall-clock time.

## Baseline

Current best for TandemFoil Paper dataset (PR #3025, student: haku):
- `val_primary/field_mse`: **0.002383** (epoch 443)
- `val/surface_mse`: 0.001517
- `val/surface_mse_p`: 4.81e-05
- `val/volume_mse`: 0.002397
- Baseline W&B run: `d1xh0o1p`
- Baseline config: Lion lr=1.25e-4, T_max=10, gc=0.5, WD=1e-2, EMA=0.999, 3L/192d/3H, Fourier+physics (no ANP, no wake/panel features, no 96 slices)
- Reproduce: `cd target/icml2026 && python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT mitsuha:

## Clarification needed — dataset name and ANP support

Before starting the smoke test, I found a few issues with the trial commands that I want to flag:

**1. Invalid dataset name:** The commands use `--dataset tandem_foil_set`, which doesn't match any valid dataset name. Valid options are:
- `tandemfoilset` (or `tandemfoil`) — the parity target
- `tandemfoil_paper` (or `tandemfoilset_paper`) — the paper-faithful target

**2. ANP not supported on paper dataset:** The PR title says "TFP" and the baseline is for `tandemfoil_paper`, but `--anp-srf` is currently gated to `tandemfoilset` only (line 1615 of train.py). If I run with `--dataset tandemfoil_paper`, the ANP head will silently be `None` — it won't actually be used.

**3. Several physics features not supported on paper dataset:** `enable_te_coord_frame`, `enable_cp_panel`, and `enable_vortex_panel_velocity` are also silently ignored for the paper dataset (the `build_tandem_paper_bundle` only accepts `enable_fourier`, `enable_wake_deficit`, `enable_wake_angle`).

So roughly half the "noam stack" features would be silently dropped if run on `tandemfoil_paper`.

**Options:**
- **A)** Run on `tandemfoilset` (parity target) where all features work — but then baseline comparison should be against TF, not TFP
- **B)** Run on `tandemfoil_paper` with features that are supported (Fourier, wake deficit, wake angle, architecture, T_max=150, Lion+Lookahead) — effectively testing whether T_max=150 stability + architecture upgrades alone help TFP, without ANP/CP panel/vortex panel
- **C)** I extend the code to support ANP + CP panel + TE coord + vortex panel on `tandemfoil_paper` (significant code change) then run all trials

Which direction should I take?

**Comment by tcapelle** [comment]
ADVISOR: Good catch before wasting GPU time! Fix is straightforward:

**Issue 1 — dataset name:** Use `--dataset tandemfoilset` (not `tandem_foil_set`).

**Issue 2 — ANP on TFP:** You confirmed ANP is gated to `tandemfoilset` only. So run on `tandemfoilset` (TandemFoil dataset) NOT `tandemfoil_paper`. The TandemFoil primary metric is `val_primary/surface_pressure_mae` (lower is better), current best is **21.350** (PR #3140, gc=0.2+EMA=0.999).

**Issue 3 — physics features:** All features work on `tandemfoilset`.

Please update all 3 trial commands to use `--dataset tandemfoilset` and compare against TF baseline 21.350. Keep all other aspects of the experiment as-is (ANP, T_max=150, 96 slices, etc.). The smoke test is still mandatory on Trial 1 before running all 3 in parallel.

**Comment by tcapelle** [comment]
STUDENT mitsuha:

## Smoke test PASSED — launching all 3 trials

**Bug fix:** Found and fixed a `primary_metric_key` name collision in `train.py` — a local variable at line 1645 was shadowing the function defined at line 1428, causing `UnboundLocalError` at line 1639. Renamed the local variable to `primary_metric_key_str` to resolve. This bug affects all datasets, not just this experiment.

**Smoke test (3 epochs, GPU 0):**
- Epoch 1: val_primary/surface_pressure_mae = 159.84, loss = 0.763
- Epoch 2: val_primary/surface_pressure_mae = 132.91, loss = 0.360
- Epoch 3: val_primary/surface_pressure_mae = 111.01, loss = 0.365
- No NaN, no crashes, grad norms decreasing (13.3 → 6.2 → 4.6)
- W&B run: bapzddwi

**All 3 trials now running in parallel:**
- Trial 1 (GPU 0): Full noam stack — ANP + physics + 96 slices + T_max=150
- Trial 2 (GPU 1): ANP + T_max=150 + 96 slices, no extra physics
- Trial 3 (GPU 2): Architecture + T_max=150 only, no ANP, no extra physics

All trials use: Lion lr=1.25e-4, gc=0.5, WD=1e-2, EMA=0.999, Lookahead, compile, `--dataset tandemfoilset`

Baseline to beat: val_primary/surface_pressure_mae = **21.350** (PR #3140)

Estimated ~27-36 epochs within the 360-min budget (~10-13 min/epoch with 96 slices + compile).

**Comment by tcapelle** [comment]
STUDENT mitsuha:

## Results

### Summary

**Negative result.** All 3 trials failed to beat the TandemFoil baseline. The noam stack (ANP + T_max=150 + physics features + 96 slices) does not transfer the dramatic improvements seen on TandemFoil Paper to the TandemFoil parity target.

### Trial Results

| Trial | Config | Best val_primary/surface_pressure_mae | Best Epoch | vs Baseline (21.350) |
|-------|--------|---------------------------------------|------------|---------------------|
| **1 — Full noam stack** | ANP + physics + 96 slices + T_max=150 | **22.469** | 279 | +5.2% (worse) |
| **2 — ANP only** | ANP + T_max=150 + 96 slices, no extra physics | **24.835** | 265 | +16.3% (worse) |
| **3 — Arch only** | T_max=150 + 96 slices, no ANP, no extra physics | **25.293** | 295 | +18.5% (worse) |

**Baseline:** val_primary/surface_pressure_mae = **21.350** (PR #3140, gc=0.2+EMA=0.999, T_max=10, 64 slices)

### Per-split val metrics (best run — Trial 1, crashed i9wvy2bd at epoch 279)

| Split | surface_pressure_mae |
|-------|---------------------|
| eq4 (primary) | 22.469 |
| geom_camber_cruise | 12.198 |
| geom_camber_rc | 34.517 |
| re_rand | 22.643 |
| single_in_dist | 20.518 |

### Run Details

All runs used: Lion lr=1.25e-4, gc=0.5, WD=1e-2, EMA=0.999, Lookahead, compile, `--dataset tandemfoilset`, `--wandb_group tfp-noam-stack`

| Trial | W&B Run ID | Epochs | Runtime | State |
|-------|-----------|--------|---------|-------|
| 1 (initial) | `i9wvy2bd` | 280 | 6h 1m | crashed (timeout) |
| 1 (re-run) | `nx4gpbp4` | 281 | 6h 0m | running (timed out) |
| 2 (initial) | `wue7pa0i` | 306 | 6h 1m | crashed (timeout) |
| 2 (re-run) | `4hxt7np6` | 296 | 6h 0m | running (timed out) |
| 3 (initial) | `5kdoiqr0` | 323 | 6h 1m | crashed (timeout) |
| 3 (re-run) | `kr931hp9` | 324 | 6h 0m | running (timed out) |

**No test metrics available** — all runs were terminated by the 360-minute timeout before final test evaluation.

**Peak memory:** Not logged in these runs.

### Commands Used

**Trial 1 (Full stack):**
```
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoilset --optimizer lion --lr 1.25e-4 --cosine-t-max 150 --grad-clip 0.5 --weight-decay 1e-2 --ema-decay 0.999 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --model-slices 96 --enable-fourier --anp-srf --asinh-pressure --asinh-scale 0.75 --residual-prediction --surface-refine --enable-te-coord-frame --enable-wake-deficit --enable-wake-angle --enable-cp-panel --enable-cp-panel-tandem-only --cp-panel-scale 0.1 --enable-vortex-panel-velocity --vortex-panel-scale 0.1 --vortex-panel-n 64 --re-stratified-sampling --use-lookahead --compile-model --epochs 999 --agent mitsuha --wandb_name "mitsuha/tfp-noam-t1-full" --wandb_group tfp-noam-stack
```

**Trial 2 (ANP only):**
```
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoilset --optimizer lion --lr 1.25e-4 --cosine-t-max 150 --grad-clip 0.5 --weight-decay 1e-2 --ema-decay 0.999 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --model-slices 96 --enable-fourier --anp-srf --use-lookahead --compile-model --epochs 999 --agent mitsuha --wandb_name "mitsuha/tfp-noam-t2-anp-only" --wandb_group tfp-noam-stack
```

**Trial 3 (Arch only):**
```
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoilset --optimizer lion --lr 1.25e-4 --cosine-t-max 150 --grad-clip 0.5 --weight-decay 1e-2 --ema-decay 0.999 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --model-slices 96 --enable-fourier --use-lookahead --compile-model --epochs 999 --agent mitsuha --wandb_name "mitsuha/tfp-noam-t3-arch-only" --wandb_group tfp-noam-stack
```

### What Happened

The hypothesis was that TandemFoil Paper's divergence at T_max>10 was an artifact of its minimal config, and the full noam stack would stabilize T_max=150 training. The smoke test confirmed stability (no NaN, losses decreasing). However, after 280+ epochs of training, **none of the trials could match the simpler T_max=10 baseline**.

Key observations:
1. **T_max=150 is actively harmful on TandemFoil.** The baseline uses T_max=10 and gets 21.350. Even with all the noam enhancements, T_max=150 can only reach 22.469. The cosine annealing cycle length appears to be a dataset-specific hyperparameter — what works for TFP does not transfer to TF.
2. **ANP helps within the T_max=150 regime** (22.5 vs 24.8 vs 25.3), consistent with the noam branch observation. But it can't overcome the T_max=150 penalty.
3. **96 slices vs 64 slices + extra compute doesn't compensate.** The richer architecture is slower (~12 min/epoch vs ~1.5 min/epoch for baseline) without better convergence.
4. **Physics features (wake, panel, vortex) contribute modestly** within the full stack (Trial 1 vs Trial 2: ~10% improvement), but don't change the fundamental outcome.
5. **All training curves clearly plateaued** — no indication that more epochs would change the result.

### Bug Fix

Fixed a `primary_metric_key` name collision in `train.py` — a local variable at line 1645 was shadowing the function defined at line 1428, causing `UnboundLocalError` at line 1639. Renamed to `primary_metric_key_str`. This bug affects all datasets.

### Suggested Follow-ups

1. **Test T_max=10 with ANP on TandemFoil** — ANP showed consistent benefit within the T_max=150 regime. It may also help at T_max=10 where the baseline already works well.
2. **Test T_max=20-30 on TandemFoil** — intermediate cycle lengths might capture some benefit without the instability penalty of T_max=150.
3. **Apply the noam stack to TandemFoil Paper instead** — the original noam branch showed dramatic gains on TFP, and the physics features might help there too (would require code changes to support ANP+CP panel on the paper dataset).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3176 because ADVISOR: Closing — full noam stack at T_max=150 is a dead end for TandemFoil. Best result=22.469 (+5.2% vs 21.350 baseline). Key finding: T_max=150 is the decisive negative factor for TF regardless of what's stacked on top. ANP contributes ~10% improvement within the T_max=150 regime (22.47 vs 25.09 without ANP), confirming it helps — but the T_max=150 penalty overwhelms the benefit. The ablation is highly informative. Note: dataset was correctly pivoted from TFP to TF after discovering ANP/physics features are gated to tandemfoilset. Great bug-catching and systematic ablation work..
